### PR TITLE
Add option to compare energy graphs with previous period

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

Add option to compare energy graphs with previous period

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/data/energy.ts [hunk 1]
```diff
@@ -350,6 +358,8 @@ const getEnergyData = async (
   }
 
   const dayDifference = differenceInDays(end || new Date(), start);
+  const period =
+    dayDifference > 35 ? "month" : dayDifference > 2 ? "day" : "hour";
```
Comment (balloob) [inline src/data/energy.ts:362]: 😬

File: src/panels/energy/ha-panel-energy.ts [hunk 2]
```diff
@@ -118,6 +136,15 @@ class PanelEnergy extends LitElement {
     };
   }
 
+  private _toggleCompare() {
+    this._compare = !this._compare;
+    const energyCollection = getEnergyDataCollection(this.hass, {
+      key: "energy_dashboard",
+    });
+    energyCollection.setCompare(this._compare);
+    energyCollection.refresh();
```
Comment (balloob) [inline src/panels/energy/ha-panel-energy.ts:145]: We only need to refresh if compare is now true and it's not loaded for current period yet?

File: src/panels/energy/ha-panel-energy.ts [hunk 3]
```diff
@@ -81,6 +86,19 @@ class PanelEnergy extends LitElement {
...
+            <ha-button-menu
+              corner="BOTTOM_START"
+              @action=${this._toggleCompare}
+            >
+              <ha-icon-button
+                .label=${this.hass.localize("panel.menu")}
+                .path=${mdiDotsVertical}
+                slot="trigger"
+              ></ha-icon-button>
+              <ha-check-list-item left .selected=${this._compare}>
```
Comment (balloob) [inline src/panels/energy/ha-panel-energy.ts:98]: Let's show this as a "comparing mode enabled" bar in the UI at the top. Probably ha-alert?

File: src/panels/energy/ha-panel-energy.ts [hunk 4]
```diff
@@ -81,6 +86,19 @@ class PanelEnergy extends LitElement {
                     collectionKey="energy_dashboard"
                   ></hui-energy-period-selector>
                 `}
+            <ha-button-menu
```
Comment (balloob) [inline src/panels/energy/ha-panel-energy.ts:89]: Let's make this a button, "COMPARE DATA" maybe as label? Or this icon 
image

**Review by balloob — DISMISSED** [review]
🎉

File: src/panels/lovelace/cards/energy/hui-energy-compare-card.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        ${formatDate(this._end || endOfDay(new Date()), this.hass.locale)}`
+          : ""}
+        with period
+        ${formatDate(this._startCompare, this.hass.locale)}${dayDifference > 0
+          ? ` -
+        ${formatDate(this._endCompare, this.hass.locale)}`
+          : ""}
+        <mwc-button slot="action" @click=${this._stopCompare}>
+          Stop comparing
+        </mwc-button>
```
Comment (balloob) [inline src/panels/lovelace/cards/energy/hui-energy-compare-card.ts:78]: I think that we can just use a X icon? Would work better on mobile too.

File: src/panels/lovelace/cards/energy/hui-energy-compare-card.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
+import { differenceInDays, endOfDay } from "date-fns";
```
Comment (balloob) [inline src/panels/lovelace/cards/energy/hui-energy-compare-card.ts:1]:
```suggestion
import { differenceInDays, endOfDay } from "date-fns/esm";
```

**Review by balloob — DISMISSED** [review]

File: src/panels/lovelace/cards/energy/hui-energy-compare-card.ts (hunk 6 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/energy/hui-energy-compare-card.ts:1]: It should already use that as it has:
```
  "main": "index.js",
  "module": "esm/index.js",
```
in `package.json`?

**Review by balloob — APPROVED** [review]

**Pull request merged**